### PR TITLE
Add popover toolbar builder that uses Flutter's new support for an iOS 16+ system toolbar (Resolves #2032)

### LINKED ISSUE #2032: [iOS] Support native edit menu + Secure Paste
Recently Flutter team worked on [[ios_edit_menu]add native edit menu](https://github.com/flutter/engine/pull/50095/files#top),
unlocking [Flutter should support iOS 15's Secure Paste feature](https://github.com/flutter/flutter/issues/103163#top)

I think SuperEditor should aim feature parity here. Currently, everytime you paste something from another app you will get a pop up. End user is annoyed.

### PR DESCRIPTION
**Pull request by matthew-carroll**
Add popover toolbar builder that uses Flutter's new support for an iOS 16+ system toolbar (Resolves #2032)

Also renamed some files for that were still using the unnecessary underscore "_" prefix.

The following code snippet should let you try the new SuperIOSTextField toolbar:

```dart
class _iOSSuperTextField extends StatefulWidget {
  const _iOSSuperTextField();

  @override
  State<_iOSSuperTextField> createState() => _iOSSuperTextFieldState();
}

class _iOSSuperTextFieldState extends State<_iOSSuperTextField> {
  final _textFieldKey = GlobalKey();
  final _focusNode = FocusNode();
  final _textController = ImeAttributedTextEditingController(
    controller: AttributedTextEditingController(
      text: AttributedText("This is some text to copy and paste."),
    ),
  );

  @override
  void dispose() {
    _textController.dispose();
    _focusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return TapRegion(
      groupId: "textfields",
      onTapOutside: (_) => _focusNode.unfocus(),
      child: TextFieldBorder(
        focusNode: _focusNode,
        borderBuilder: _borderBuilder,
        child: SuperIOSTextField(
          key: _textFieldKey,
          focusNode: _focusNode,
          tapRegionGroupId: "textfields",
          textController: _textController,
          textStyleBuilder: _textStyleBuilder,
          hintBuilder: _createHintBuilder("Enter single line text..."),
          padding: const EdgeInsets.all(4),
          minLines: 1,
          maxLines: 1,
          caretStyle: CaretStyle().copyWith(color: Colors.red),
          selectionColor: Colors.red,
          handlesColor: Colors.red,
          popoverToolbarBuilder: iOSSystemPopoverToolbarWithBackupFlutterVersion,
        ),
      ),
    );
  }
}
```

https://github.com/user-attachments/assets/76224c0c-c479-4574-951e-f9c4e5b7a201

### CODE REVIEW

**Comment by matthew-carroll** [comment]
@BazinC - I added iOS system menu support to `SuperIOSTextfield` - would you like to try that out on the text field before I add it to `SuperEditor`?

File: super_editor/lib/src/super_textfield/ios/ios_textfield.dart [hunk 1]
```diff
@@ -694,7 +697,23 @@ class SuperIOSTextFieldState extends State<SuperIOSTextField>
   }
 }
 
-Widget _defaultPopoverToolbarBuilder(BuildContext context, IOSEditingOverlayController controller) {
+/// Builder that returns a widget for an iOS-style popover editing toolbar.
+typedef IOSPopoverToolbarBuilder = Widget Function(BuildContext, IOSEditingOverlayController);
+
+/// An [IOSPopoverToolbarBuilder] that displays the iOS system popover toolbar, if the version of
+/// iOS is recent enough, otherwise builds [defaultIosPopoverToolbarBuilder].
+Widget iOSSystemPopoverToolbarWithBackupFlutterVersion(BuildContext context, IOSEditingOverlayController controller) {
```
Comment (angelosilvestre) [inline super_editor/lib/src/super_textfield/ios/ios_textfield.dart:705]: Is `WithBackupFlutterVersion` an appropriate name? I think  `iOSSystemPopoverToolbarWithFallback` would be simpler.

**Review by angelosilvestre — APPROVED** [review]
LGTM with a single comment.

**Comment by BazinC** [comment]
> @BazinC - I added iOS system menu support to `SuperIOSTextfield` - would you like to try that out on the text field before I add it to `SuperEditor`?

@matthew-carroll We don't use SuperIOSText. Is this PR related to the following comment? https://github.com/superlistapp/super_editor/issues/2032#issuecomment-2132136212
Is it still pending on Flutter side?

**Comment by matthew-carroll** [comment]
@BazinC this PR adds the use of the native menu. It adds it to SuperTextField, first, because that's simpler. Can you please try out a SuperTextField on iOS and check if the native menu meets your expectations?

If we need to make changes to the behavior, or if the menu won't work, then it's less expensive to discover that here and now.

**Comment by BazinC** [comment]
@matthew-carroll can you tell me which Flutter version you used for this branch?
SystemContextMenu is only available on master and I fail to run the demo on master
<details>
<summary>logs</summary>

```
corentinbazin@macbook-pro example % flutter run lib/flutter_demos/main_flutter_textfield.dart -d 00008110-001229512698401E
Launching lib/flutter_demos/main_flutter_textfield.dart on iPhone in debug mode...
Upgrading AppFrameworkInfo.plist
Upgrading Podfile
Upgrading Info.plist
ios/Runner/AppDelegate.swift uses the deprecated @UIApplicationMain attribute, updating.
Automatically signing iOS for device deployment using specified development team in Xcode project: 5RJWFAUGXQ
Running pod install...                                             807ms
Running Xcode build...                                                  
Xcode build done.                                           13.9s
Failed to build iOS app
Could not build the precompiled application for the device.
Error (Xcode): ../../../../../.pub-cache/hosted/pub.dev/win32-5.0.9/lib/src/guid.dart:32:9: Error: Type 'UnmodifiableUint8ListView' not found.



Error launching application on iPhone.
corentinbazin@macbook-pro example % flutter doctor                                                                        
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 3.24.0-1.0.pre.54, on macOS 14.5 23F79 darwin-arm64, locale en-FR)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.1)
[✓] VS Code (version 1.91.0)
[✓] Connected device (5 available)
    ! Error: Browsing on the local area network for iPhone de Alois. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for Corentin’s Apple Watch. Ensure the device is unlocked and discoverable via Bluetooth. (code -27)
[✓] Network resources

• No issues found!
```
</details>

Tried again after cleaning pub_cache, with no luck

**Comment by matthew-carroll** [comment]
@BazinC I brought in the latest changes from `main` in case that was causing an issue.

As for the Flutter branch - I just tried on Beta and the context menu works, so you don't need to be on `master`.

For the demo - run the "Example - All Demos" and then choose the iOS super text field from the navigation drawer. When you open the editing toolbar in that demo, it will be the native iOS edit menu.

**Comment by BazinC** [comment]
> @BazinC I brought in the latest changes from `main` in case that was causing an issue.
> 
> As for the Flutter branch - I just tried on Beta and the context menu works, so you don't need to be on `master`.
> 
> For the demo - run the "Example - All Demos" and then choose the iOS super text field from the navigation drawer. When you open the editing toolbar in that demo, it will be the native iOS edit menu.

Thanks! I'm trying with beta now and it works! 
I don't see any bug at first glance. No more "Allow/Don't Allow Paste" dialog.
One question: How customizable it could be? Now I see Cut/Copy/Paste/Delete/Select All/Autofill.
Not a big deal for now.

Toolbar position and its arrow position looks off but it's not part of this ticket.

**Comment by matthew-carroll** [comment]
@angelosilvestre do you wanna take another look at this PR? I extended the toolbar work to include Super Editor. I think we're good to go at this point.

File: super_editor/example/lib/demos/in_the_lab/feature_ios_native_context_menu.dart [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+class _SuperTextFieldWithNativeContextMenuState extends State<_SuperTextFieldWithNativeContextMenu> {
+  @override
+  Widget build(BuildContext context) {
+    return DecoratedBox(
+      decoration: BoxDecoration(
+        border: Border.all(color: Colors.grey),
+        borderRadius: BorderRadius.circular(4),
+        // color: const Color(0xFF222222),
```
Comment (angelosilvestre) [inline super_editor/example/lib/demos/in_the_lab/feature_ios_native_context_menu.dart:159]: Please remove the commented code.

**Review by angelosilvestre — APPROVED** [review]
LGTM

**Comment by ibrahimdevs** [comment]
@matthew-carroll , @angelosilvestre this is well-needed improvement, thanks. But as far as I understand, we can only use it via SuperIOSTextField. 

Do you think to put it SuperTextField as default? 
Or any variable to control popoverToolbarBuilder on SuperTextField? Maybe preferSystemToolbar: true/false

**Comment by matthew-carroll** [comment]
@ibrahimdevs this Flutter feature is specifically for iOS. We don't know when/if Flutter will add the same capability for Android.

If they already have, feel free to point us to the Flutter PR, or docs with that info. Thanks.

**Comment by ibrahimdevs** [comment]
@matthew-carroll I think I can't explain well. SuperIOSTextField's default popoverToolbarBuilder should be this one.

```
class SuperIOSTextField extends StatefulWidget {
  const SuperIOSTextField({
    Key? key,
    this.focusNode,
    this.tapRegionGroupId,
    this.textController,
    this.textStyleBuilder = defaultTextFieldStyleBuilder,
    this.textAlign = TextAlign.left,
    this.padding,
    this.hintBehavior = HintBehavior.displayHintUntilFocus,
    this.hintBuilder,
    this.minLines,
    this.maxLines = 1,
    this.lineHeight,
    required this.caretStyle,
    this.blinkTimingMode = BlinkTimingMode.ticker,
    required this.selectionColor,
    required this.handlesColor,
    this.textInputAction,
    this.imeConfiguration,
    this.showComposingUnderline = true,
    this.popoverToolbarBuilder = iOSSystemPopoverTextFieldToolbarWithFallback,
    // instead of => this.popoverToolbarBuilder = defaultIosPopoverToolbarBuilder,
    this.showDebugPaint = false,
  }) : super(key: key);
```

I'm using SuperTextField and It uses [SuperDesktopTextField], [SuperAndroidTextField] and [SuperIOSTextField] under the hood. But there is no way to change/use iOSSystemPopoverTextFieldToolbarWithFallback with default SuperTextField implementation.

**Comment by matthew-carroll** [comment]
@angelosilvestre do we not have the native context menu in the ancestor controls widget?

**Comment by angelosilvestre** [comment]
@matthew-carroll We don't currently have a way to use the native context menu for text fields without using `SuperIOSTextField`. On `SuperEditor` we have the `SuperEditorIosControlsScope` to do that. But we don't have a similar way to do it in `SuperTextField`.

**Comment by matthew-carroll** [comment]
@angelosilvestre can you file a ticket that explains the state of things and request some kind of configuration support to get the native context menu in a regular `SuperTextField`?

**Comment by ibrahimdevs** [comment]
For anyone needs this progress, you can check: https://github.com/superlistapp/super_editor/issues/2550

**Pull request merged**